dash/validator_conn_executor.go

```diff
@@ -0,0 +1,321 @@
+const validatorConnExecutorName = "ValidatorConnExecutor"
+
+// iSwitch defines p2p.Switch methods that are used by this Executor.
+// Useful to create a mock  of the p2p.Switch.
+type iSwitch interface {
+	Peers() p2p.IPeerSet
+
+	AddPersistentPeers(addrs []string) error
+	RemovePersistentPeer(addr string) error
+
+	DialPeersAsync(addrs []string) error
+	IsDialingOrExistingAddress(*p2p.NetAddress) bool
+	StopPeerGracefully(p2p.Peer)
+}
+
+// ValidatorConnExecutor retrieves validator update events and establishes new validator connections
+// within the ValidatorSet.
+// If it's already connected to a member of current validator set, it will keep that connection.
+// Otherwise, it will randomly select some members of the active validator set and connect to them, to ensure
+// it has connectivity with at least `NumConnections` members of the active validator set.
+//
+// Note that we mark peers that are members of active validator set as Persistent, so p2p subsystem
+// will retry the connection if it fails.
+type ValidatorConnExecutor struct {
+	service.BaseService
+	ctx          context.Context
+	nodeID       p2p.ID
+	eventBus     *types.EventBus
+	p2pSwitch    iSwitch
+	subscription types.Subscription
+
+	// validatorSetMembers contains validators active in the current Validator Set, indexed by node ID
+	validatorSetMembers validatorMap
+	// connectedValidators contains validators we should be connected to, indexed by node ID
+	connectedValidators validatorMap
+
+	// mux is a mutex to ensure only one goroutine is processing connections
+	mux sync.Mutex
+
+	// *** configuration *** //
+
+	// NumConnections is the number of connections to establish, defaults to 3
+	NumConnections int
+	// EventBusCapacity sets event bus buffer capacity, defaults to 10
+	EventBusCapacity int
+}
+
+// NewValidatorConnExecutor creates a Service that connects to other validators within the same Validator Set.
+// Don't forget to Start() and Stop() the service.
+func NewValidatorConnExecutor(
+	nodeID p2p.ID,
+	eventBus *types.EventBus,
+	sw iSwitch,
+	logger log.Logger) *ValidatorConnExecutor {
+	vc := &ValidatorConnExecutor{
+		ctx:                 context.Background(),
+		nodeID:              nodeID,
+		eventBus:            eventBus,
+		p2pSwitch:           sw,
+		NumConnections:      3,
+		EventBusCapacity:    10,
+		validatorSetMembers: validatorMap{},
+		connectedValidators: validatorMap{},
+	}
+
+	baseService := service.NewBaseService(logger, validatorConnExecutorName, vc)
+	vc.BaseService = *baseService
+
+	return vc
+}
+
+// OnStart implements Service to subscribe to Validator Update events
+func (vc *ValidatorConnExecutor) OnStart() error {
+	if err := vc.subscribe(); err != nil {
+		return err
+	}
+
+	go func() {
+		var err error
+		for err == nil {
+			err = vc.receiveEvents()
+		}
+		vc.Logger.Error("ValidatorConnExecutor goroutine finished", "reason", err)
+	}()
```

Comment:
let's consider the following implementation, it allows us to eliminate a dependency on a context in a structure:
```
go func() {
	var err error
	for err == nil {
		ctx := context.Background()
		// here you can adjust the context if it is needed
		// for instance, to make a timeout for handling of an event
		ctx, cancel := context.WithTimeout(ctx, 5 * time.Second)
		err = vc.receiveEvents(ctx) // also, you need to add a context as an argument
		cancel()
	}
	vc.Logger.Error("ValidatorConnExecutor goroutine finished", "reason", err)
}()
```